pkg/cache/context.go

```diff
@@ -499,10 +499,49 @@ func (ctx *Context) GetApplication(appID string) interfaces.ManagedApp {
 func (ctx *Context) RemoveApplication(appID string) error {
 	ctx.lock.Lock()
 	defer ctx.lock.Unlock()
-	if _, exist := ctx.applications[appID]; exist {
+	if app, exist := ctx.applications[appID]; exist {
```

Comment:
This can be simplified. We can group a set of states into a group, e.g **Terminated**. That includes state: `Rejected`, `Killed`, `Failed`, `Completed`. If we found any task whose state is not in this group, then we fail with some error. I think you can do something similar to https://github.com/apache/incubator-yunikorn-k8shim/blob/1c46df413f290941616a4cb701e6ddd291851667/pkg/common/events/states.go#L75.